### ISSUE TITLE
Gradients not preserved by IAF caching [bug]

### ISSUE THREAD

**Issue by adam-coogan** [pyro-ppl/pyro#1885]
The `InverseAutoregressiveFlow` class does not preserve gradient information when computing `log_prob` using cached values.  Here is a minimal example:
```python
from pyro import distributions as dist
from pyro.distributions import InverseAutoregressiveFlow as IAF
from pyro.nn import AutoRegressiveNN
import torch

base_dist = dist.Normal(torch.zeros(2), torch.ones(2))
iaf_transform = IAF(AutoRegressiveNN(2, [4]))
p_iaf  = dist.TransformedDistribution(base_dist, iaf_transform)
```
Computing the density for an arbitrary point gives gradient information:
```python
>>> p_iaf.log_prob(torch.randn(2))
    tensor(-3.0375, grad_fn=<AddBackward0>)
```
That is not the case for a point sampled from the IAF, for which the log probability is cached:
```python
>>> x = p_iaf.sample()
>>> p_iaf.log_prob(x)
    tensor(-2.4662)
```

It's not obvious to me how to fix this. I'm on the dev branch of pyro, and can provide more information if necessary.

**Comment by martinjankowiak** [comment]
you need to use `rsample` if you want gradients.

please reopen if that doesn't fix your issue.
